### PR TITLE
Fix beta and exponential sigmas + add tests

### LINKED ISSUE #9951: All schedulers have broken beta and exponential sigma methods
### Describe the bug

following schedulers implement beta and exponential sigma methods:
```log
scheduling_deis_multistep.py
scheduling_dpmsolver_multistep.py
scheduling_dpmsolver_multistep_inverse.py
scheduling_dpmsolver_sde.py
scheduling_dpmsolver_singlestep.py
scheduling_euler_discrete.py
scheduling_heun_discrete.py
scheduling_k_dpm_2_ancestral_discrete.py
scheduling_k_dpm_2_discrete.py
scheduling_lms_discrete.py
scheduling_sasolver.py
scheduling_unipc_multistep.py
```

and in all of them, its pretty much broken.
(ok, i admit i tested only 6, not all)

looking at code segment:
```py
        elif self.config.use_exponential_sigmas:
            sigmas = self._convert_to_exponential(in_sigmas=sigmas, num_inference_steps=self.num_inference_steps)
            timesteps = np.array([self._sigma_to_t(sigma, log_sigmas) for sigma in sigmas])
        elif self.config.use_beta_sigmas:
            sigmas = self._convert_to_beta(in_sigmas=sigmas, num_inference_steps=self.num_inference_steps)
            timesteps = np.array([self._sigma_to_t(sigma, log_sigmas) for sigma in sigmas])
```
1. it refers to `log_sigmas`, but they are only set if `self.config.use_karras_sigmas`, otherwise its UNDEFINED
2. it refers to `self.num_inference_steps`, but that param is set AFTER this code segment and here its NONE
3. actual sigmas at the end are blindly converted from numpy to tensor even if they are already a tensor causing a runtime error

### Reproduction

simply use ANY of the above schedulers with either `use_exponential_sigmas` or `use_beta_sigmas`

### Logs

```shell
UnboundLocalError: cannot access local variable 'log_sigmas' where it is not associated with a value

or 

TypeError: linspace() received an invalid combination of arguments - got (float, float, NoneType), but expected one of:

or 

TypeError: 'NoneType' object cannot be interpreted as an integer

or 

TypeError: expected np.ndarray (got Tensor)
```

### System Info

diffusers from main branch as of 11/18/24 commit #345907f

### Who can help?

@yiyixuxu @sayakpaul @DN6 @asomoza

### PR DESCRIPTION
**Pull request by hlky**
# What does this PR do?

- `log_sigmas` moved out of `karras` branch
- `_convert_to_exponential` and `_convert_to_beta` return `np.ndarray` for consistency with `_convert_to_karras` and other code
- Just use `num_inference_steps` for input to `_convert_to_exponential` and `_convert_to_beta`, not all schedulers update `self.num_inference_steps`

Fixes #9951

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu @vladmandic

### CODE REVIEW

**Comment by vladmandic** [comment]
that was fast! :)

**Comment by sayakpaul** [comment]
@vladmandic could you confirm if this is working for you? 

@hlky this is extraordinarily fast! Thank you so much!

**Review by yiyixuxu — APPROVED** [review]
thanks for the fix!

**Comment by hlky** [comment]
Also fixes incorrect ordering in some schedulers.

**Comment by vladmandic** [comment]
LGTM!
image
image (1)

**Comment by sayakpaul** [comment]
That was a rigorous test, thanks! 

Wonder what did you use for the collage? :D

**Comment by vladmandic** [comment]
> Wonder what did you use for the collage? :D

sdnext built-in xyz grid functionality

image

File: src/diffusers/schedulers/scheduling_unipc_multistep.py [hunk 1]
```diff
@@ -347,11 +347,33 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
                 )
             sigmas = np.concatenate([sigmas, [sigma_last]]).astype(np.float32)
         elif self.config.use_exponential_sigmas:
-            sigmas = self._convert_to_exponential(in_sigmas=sigmas, num_inference_steps=self.num_inference_steps)
+            log_sigmas = np.log(sigmas)
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_unipc_multistep.py:350]: ohh so it looked like it was wrong before, thank!

**Review by yiyixuxu — COMMENTED** [review]
thanks @hlky!
looked like there were some issues that you fixed here that did not get caught in our previous tests or @vladmandic 's test, I will try to make & run a more complete slow-test on my end and then merge

**Pull request merged**